tracer/src/Datadog.Trace/Baggage.cs

```diff
@@ -0,0 +1,357 @@
+﻿// <copyright file="Baggage.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+
+#nullable enable
+
+namespace Datadog.Trace;
+
+/// <summary>
+/// Baggage is a collection of name-value pairs that are propagated to downstream services.
+/// </summary>
+internal class Baggage : IDictionary<string, string>
+{
+    private static readonly AsyncLocal<Baggage> AsyncStorage = new();
+
+    private static readonly List<KeyValuePair<string, string>> EmptyList = [];
+
+    private List<KeyValuePair<string, string>>? _items;
```

Comment:
Can we use a `Dictionary<string,string>?` instead of a `List<KeyValuePair<string, string>>?` ? From a glance it seems like it may be easier to maintain

Reply:
Maybe a case of premature optimization, but in the common use case users will add very few baggage items (2-5 ish, probably 10 max). For such small collections, `List<T>` can be more efficient than a dictionary, and also leaner (less heap allocations).

I did benchmarks on this back when we added `TraceTagCollection` which has similar usage:
- small number of items
- very few `add` operations
- few `get` operation by key, if any
- collection iterated when propagating (not using `get`)

Reply:
Gotcha, well I'm happy to keep it so long as the "public" API is unaware of the implementation detail!